contracts/HoprStake2.sol

```diff
@@ -0,0 +1,443 @@
+     * @param operator address operator requesting the transfer
+     * @param from address token holder address
+     * @param tokenId uint256 amount of tokens to transfer
+     * @param data bytes hex information provided by the token holder (if any)
+     */
+    function onERC721Received(
+        // solhint-disable-next-line no-unused-vars
+        address operator,
+        address from,
+        uint256 tokenId,
+        // solhint-disable-next-line no-unused-vars
+        bytes calldata data
+    ) external override returns (bytes4) {
+        require(_msgSender() == address(NFT_CONTRACT), "HoprStake: Cannot SafeTransferFrom tokens other than HoprBoost.");
+        require(block.timestamp <= PROGRAM_END, "HoprStake: Program ended, cannot redeem boosts.");
+        // Account memory account = accounts[from];
+        _sync(from);
+
+        // redeem NFT
+        redeemedNft[from][redeemedNftIndex[from]] = tokenId;
+        redeemedNftIndex[from] += 1;
+
+        // update boost factor
+        uint256 typeId = NFT_CONTRACT.typeIndexOf(tokenId);
+        (uint256 factor, ) = NFT_CONTRACT.boostOf(tokenId);
+
+        uint256 boostIndex = redeemedFactorIndex[from];
+        uint256 index = 0;
+        for (index; index < boostIndex; index++) {
+            // loop through redeemed factors, replace the factor of the same type, if the current factor is larger.
+            uint256 redeemedId = redeemedFactor[from][index];
+            (uint256 redeemedFactorValue, ) = NFT_CONTRACT.boostOf(redeemedId);
+
+            if (NFT_CONTRACT.typeIndexOf(redeemedId) == typeId) {
+                if (redeemedFactorValue < factor) {
+                    redeemedFactor[from][index] = tokenId;
+                }
+                emit Redeemed(from, tokenId, redeemedFactorValue < factor);
+                break;
+            }
+        }
+        if (index == boostIndex) {
+            // new type being redeemed.
+            redeemedFactor[from][boostIndex] = tokenId;
+            redeemedFactorIndex[from] += 1;
+            emit Redeemed(from, tokenId, true);
+        }
+
+        return IERC721Receiver(address(this)).onERC721Received.selector;
+    }
+
+    /**
+     * @dev Manually sync account's reward states
+     * @notice public function of ``_sync``.
+     * @param account address Account whose stake rewards will be synced.
+     */
+    function sync(address account) external {
+        _sync(account); 
+    }
+
+    /**
+     * @dev Sync rewards and claim them
+     * @notice public function of ``_sync`` + ``_claim``
+     * @param account address Account whose stake rewards will be synced and claimed.
+     */
+    function claimRewards(address account) external {
+        _sync(account); 
+        _claim(account);
+    }
+
+    /**
+     * @dev Unlock staking
+     * @param account address Account that staked tokens.
+     */
+    function unlock(address account) external {
```

Comment:
Would it make sense to also expose an `unlock()` function that passes `msg.sender` to make sure there's less stuff that can go wrong? I see 99.9% of all users being fine with that.
Probably relevant for the UI implementation. CC @nionis

Reply:
I can make this function `unlockFor` and make another `unlock()` for `msg.sender`

Reply:
sgtm @QYuQianchen